src/eg/src/verifiable_decryption.rs

```diff
@@ -0,0 +1,717 @@
+    pub i: GuardianIndex,
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    /// The decryption share value
+    pub m_i: BigUint,
+}
+
+impl DecryptionShare {
+    /// This function computes the [`DecryptionShare`] for a given [`Ciphertext`] and [`GuardianSecretKeyShare`].
+    /// The arguments are
+    /// - self - the encrypted share
+    /// - fixed_parameters - the fixed parameters
+    /// - secret_key_share - the secret key share
+    /// - ciphertext - the ElGamal ciphertext
+    pub fn from(
+        fixed_parameters: &FixedParameters,
+        secret_key_share: &GuardianSecretKeyShare,
+        ciphertext: &Ciphertext,
+    ) -> Self {
+        let m_i = ciphertext
+            .alpha
+            .modpow(&secret_key_share.p_i, fixed_parameters.p.as_ref());
+        DecryptionShare {
+            i: secret_key_share.i,
+            m_i,
+        }
+    }
+}
+
+/// The combined decryption share allows to compute the plain-text from a given ciphertext. This corresponds to the `M` in Section `3.6.2`.
+pub struct CombinedDecryptionShare(BigUint);
+
+/// Represents errors occurring while combining [`DecryptionShare`]s into a [`CombinedDecryptionShare`].
+#[derive(Error, Debug)]
+pub enum ShareCombinationError {
+    /// Occurs if not enough shares were provided. Combination requires shares from at least `k` out of `n` guardians.
+    #[error("Only {l} decryption shares given, but at least {k} required.")]
+    NotEnoughShares { l: usize, k: u32 },
+    /// Occurs if multiple shares from the same guardian are provided.
+    #[error("Guardian {i} is represented more than once in the decryption shares.")]
+    DuplicateGuardian { i: GuardianIndex },
+}
+
+impl CombinedDecryptionShare {
+    /// This functions combines decryption shares
+    /// The arguments are
+    /// - election_parameters - the election parameters
+    /// - decryption_shares - a vector of decryption shares
+    /// The computation follows Equation `68`.
+    pub fn combine(
+        election_parameters: &ElectionParameters,
+        decryption_shares: &[DecryptionShare],
+    ) -> Result<Self, ShareCombinationError> {
+        let fixed_parameters = &election_parameters.fixed_parameters;
+        let varying_parameters = &election_parameters.varying_parameters;
+        let n = varying_parameters.n.get_one_based_usize();
+        let k = varying_parameters.k.get_one_based_u32();
+
+        let l = decryption_shares.len();
+        // Ensure that we have at least k decryption shares
+        if l < k as usize {
+            return Err(ShareCombinationError::NotEnoughShares { l, k });
+        }
+        // Verify that every guardian (index) is represented at most once.
+        let mut seen = vec![false; n];
+        for share in decryption_shares {
+            let seen_ix = share.i.get_zero_based_usize();
+            if seen[seen_ix] {
```

Comment:
Are we sure this always exists, as in, seen_ix < n?

Reply:
Good point! I have added a check to ensure that the indices are within the bounds.